File: src/main/java/com/kigya/patterns/java/strategy/entity/DecoyDuck.java
package com.kigya.patterns.java.strategy.entity;

import com.kigya.patterns.java.strategy.implementation.FlyNoWay;
import com.kigya.patterns.java.strategy.implementation.MuteQuack;
import lombok.extern.flogger.Flogger;

import java.util.logging.Level;

@Flogger
public class DecoyDuck extends Duck {

    public DecoyDuck() {
        setFlyBehaviour(new FlyNoWay());
        setQuackBehaviour(new MuteQuack());
        display();
        performQuack();
        performFly();
    }

    public void display() {
        log.at(Level.INFO).log("I'm a duck Decoy", DecoyDuck.class.toString());
    }
}

File: src/main/java/com/kigya/patterns/kotlin/strategy/entity/DecoyDuck.kt
package com.kigya.patterns.kotlin.strategy.entity

data class DecoyDuck(
    override var quackBehaviour: Quack = ::flyNoWay,
    override var flyBehaviour: Fly = ::muteQuack
) : Duck() {

    override fun display(): Unit = logger.info {
        "I'm a duck Decoy: " + DecoyDuck::class.simpleName.toString()
    }
}